pkg/storage/stores/shipper/compactor/table.go

```diff
@@ -69,105 +79,118 @@ func newTable(ctx context.Context, workingDirectory string, indexStorageClient s
 		name:               filepath.Base(workingDirectory),
 		workingDirectory:   workingDirectory,
 		indexStorageClient: indexStorageClient,
-		quit:               make(chan struct{}),
-		applyRetention:     applyRetention,
 		tableMarker:        tableMarker,
+		expirationChecker:  expirationChecker,
+		indexSets:          map[string]*indexSet{},
+		baseUserIndexSet:   storage.NewIndexSet(indexStorageClient, true),
+		baseCommonIndexSet: storage.NewIndexSet(indexStorageClient, false),
+		seedSourceFileIdx:  -1,
 	}
 	table.logger = log.With(util_log.Logger, "table-name", table.name)
 
 	return &table, nil
 }
 
-func (t *table) compact(tableHasExpiredStreams bool) error {
-	indexFiles, err := t.indexStorageClient.ListFiles(t.ctx, t.name)
+func (t *table) compact(applyRetention bool) error {
```

Comment:
Can we separate the logic for multi tenant index, this function is quite hard to understand

Reply:
I did give it a try but it looks hard. I feel the complexity comes from the support for backwards compatibility and multiple index formats. I will see if I can do something or maybe we can discuss this in a call face to face.

Reply:
> LGTM it's a bit difficult to follow.
> 
> With the PR description I was able to understand everything though. Although we might want to think about refactoring to make the code cleaner, single code path for each case and separate concern. Talking specifically about the compaction/retention code.

I did some thinking and it does not seem as easy to build a separate execution path without making it complex. I am going to merge this PR and open a separate PR if I can come up with something.

Thanks for quickly reviewing this PR!